### PR TITLE
DrivAerML: 5-epoch linear warmup + cosine T_max=30

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The DrivAerML champion uses cosine annealing starting from full LR at epoch 0. A linear warmup ramp in early epochs could prevent cold-start instability and find a better initial basin. 5-epoch warmup is gentle enough to not waste budget.

## Instructions

Add a 5-epoch linear warmup before cosine annealing. Implement as follows:
- For epochs 0-4: use `lr = base_lr * (epoch + 1) / 5` (linear ramp from lr/5 to lr)
- For epochs 5+: standard cosine annealing with T_max=30
- Simplest implementation: `torch.optim.lr_scheduler.LinearLR(optimizer, start_factor=0.2, end_factor=1.0, total_iters=5)` for first 5 epochs, then switch to `CosineAnnealingLR(optimizer, T_max=30)`
- Or use `SequentialLR([LinearLR(...), CosineAnnealingLR(...)], milestones=[5])`
- Add a `--warmup-epochs` flag (default 0)

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --warmup-epochs 5 \
  --wandb_group dm-warmup-sweep
```

## Reporting Requirements
- Best val `surface_rel_l2_pct` + epoch
- Test `surface_rel_l2_pct` at best val checkpoint
- W&B run ID

## Baseline

**DrivAerML baselines:**
- val best: `3.997%` @ ep467 (PR #2898, W&B: `bht6h42t`, 4L/512d, AdamW lr=5e-4, T_max=30)
- test best: `6.244%` (PR #2648, 4L/320d, truncated eval — OLDER CONFIG)
- External targets: AB-UPT 3.82% | Transolver-3: 3.71%

```bash
# DrivAerML champion (val 3.997% @ ep467)
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT chopper:

## Results

### DrivAerML: 5-epoch linear warmup + cosine T_max=30

| Metric | Baseline | Warmup (this run) |
|---|---|---|
| val `surface_rel_l2_pct` (best) | 3.997% @ ep467 | **5.823% @ ep163** |
| test `surface_rel_l2_pct` (at best val) | 6.244% | **7.338%** |
| External targets | AB-UPT 3.82% / Transolver-3 3.71% | — |

**W&B run:** `ny2kvo8o` — [link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/ny2kvo8o)

**Command:**
```bash
CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --warmup-epochs 5 \
  --agent chopper --wandb-name "chopper/dm-linear-warmup-5ep" \
  --wandb-group dm-warmup-sweep
```

**Peak memory:** N/A (no memory metric logged; GPU memory freed after run)

### What happened

**Negative result.** The 5-epoch linear warmup (lr/5 → lr over 1970 optimizer steps, then per-step cosine T_max=30) performed significantly worse than the baseline:

- Best val 5.823% vs baseline 3.997% — **1.83pp worse**
- The model reached its best val at epoch 163, then diverged catastrophically. By epoch 523, val was 63.9% with grad norms exploding to 6.6 billion.

The warmup ramp correctly went lr=1e-4 → 5e-4 over epochs 1-5, then handed off to cosine annealing. The early convergence was reasonable (30.5% at ep5, 24.2% at ep10) but the model never reached the quality of the no-warmup baseline. The hypothesis that cold-start instability was hurting the champion config appears false — the champion's immediate full-LR start into per-step cosine cycling may actually provide beneficial exploration in early epochs that the gentle warmup ramp removes.

Note on implementation: the scheduler steps per optimizer step (not per epoch), consistent with the existing CosineAnnealingLR setup. warmup_epochs=5 × 394 steps/epoch = 1970 warmup steps with LinearLR(start_factor=0.2).

### Suggested follow-ups

- The late-training divergence (post ep163) is concerning and may indicate the per-step cosine cycling becomes pathological over long runs. Could investigate combining warmup with gradient clipping or EMA to stabilize late training.
- A much shorter warmup (1-2 epochs) might warm up just enough without dulling early exploration, though given the 1.8pp gap it's unlikely to close the deficit.
- The champion config's per-step T_max=30 cosine may inherently act as a "warm restart" mechanism that makes explicit warmup redundant.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3069 because ADVISOR: Closing as a clear negative result. W&B run ny2kvo8o confirms: best val 5.823% @ ep163 vs baseline 3.997% — 1.83pp worse. The run then diverged catastrophically (grad norm reaching 3×10^10, val=63.9% by ep523).

The warmup hypothesis is falsified. The 5-ep warmup variant (this PR) and the 10-ep warmup variant (PR #3070, best val 6.225%, also closed) are both substantially worse than the no-warmup champion. The champion's immediate full-LR cosine cycling is not a cold-start liability — it provides beneficial early exploration that the gentle warmup ramp suppresses.

The underlying issue is late-training divergence without gradient clipping, not cold-start instability. That mechanism is already being addressed in the EMA+gc thread (PR #3136). No further warmup variants are warranted. Thank you for the thorough analysis and the accurate diagnosis in your suggested follow-ups..
